pkg/agent/flowexporter/connections/connections.go

```diff
@@ -28,42 +27,64 @@ import (
+	IterateCxnMapWithCB(updateCallback flowexporter.FlowRecordUpdate) error
+	FlushConnectionStore()
 }
 
 type connectionStore struct {
-	connections map[flowexporter.ConnectionKey]flowexporter.Connection // Add 5-tuple as string array
-	connDumper  ConnTrackDumper
-	ifaceStore  interfacestore.InterfaceStore
-	mutex       sync.Mutex
+	// pollDone channel is used for synchronization of poll(ConnectionStore.Run) and export(FlowExporter.Run) go routines.
+	// Therefore, there is no requirement of lock to make connections map thread safe.
+	connections    map[flowexporter.ConnectionKey]flowexporter.Connection // Add 5-tuple as string array
+	connDumper     ConnTrackDumper
+	ifaceStore     interfacestore.InterfaceStore
+	pollInterval   time.Duration
+	exportInterval time.Duration
 }
 
-func NewConnectionStore(ctDumper ConnTrackDumper, ifaceStore interfacestore.InterfaceStore) *connectionStore {
+func NewConnectionStore(ctDumper ConnTrackDumper, ifaceStore interfacestore.InterfaceStore, pollInterval time.Duration, exportInterval time.Duration) *connectionStore {
 	return &connectionStore{
-		connections: make(map[flowexporter.ConnectionKey]flowexporter.Connection),
-		connDumper:  ctDumper,
-		ifaceStore:  ifaceStore,
+		connections:    make(map[flowexporter.ConnectionKey]flowexporter.Connection),
+		connDumper:     ctDumper,
+		ifaceStore:     ifaceStore,
+		pollInterval:   pollInterval,
+		exportInterval: exportInterval,
 	}
 }
 
 // Run polls the connTrackDumper module periodically to get connections. These connections are used
 // to build connection store.
-func (cs *connectionStore) Run(stopCh <-chan struct{}) {
+func (cs *connectionStore) Run(stopCh <-chan struct{}, pollDone chan bool) {
 	klog.Infof("Starting conntrack polling")
 
-	ticker := time.NewTicker(flowexporter.PollInterval)
-	defer ticker.Stop()
+	pollTicker := time.NewTicker(cs.pollInterval)
+	defer pollTicker.Stop()
+	exportTrigger := uint(cs.exportInterval / cs.pollInterval)
+	exportCounter := uint(0)
+
 	for {
 		select {
 		case <-stopCh:
 			break
-		case <-ticker.C:
+		case <-pollTicker.C:
+			if exportCounter == 0 {
+				exportCounter = exportTrigger
+				// Flush connection map once all flow records are sent.
+				// TODO: Optimize this logic by flushing individual connections based on the individual timeout values.
+				cs.FlushConnectionStore()
+			}
 			_, err := cs.poll()
 			if err != nil {
 				// Not failing here as errors can be transient and could be resolved in future poll cycles.
 				// TODO: Come up with a backoff/retry mechanism by increasing poll interval and adding retry timeout
 				klog.Errorf("Error during conntrack poll cycle: %v", err)
 			}
+			exportCounter = exportCounter - 1
+			if exportCounter == 0 {
+				// We need synchronization between ConnectionStore.Run and FlowExporter.Run go routines. pollDone channel provides that synchronization.
+				// More details in exporter/exporter.go.
+				pollDone <- true
```

Comment:
reading the code, it seems that the name is not appropriate. You do not send a signal when polling is done, you send a signal when the exporter needs to run.

Reply:
what you can do is send a notification on the channel for every poll, and let the exporter decide when to export based on the number of signals received. I think that's a bit more elegant.

Reply:
That's a good suggestion and I think it is well use of channel properties (I am a newbie to go channels and still exploring :)). 
I was thinking of using a separate mutex for this instead of channel for the purpose of starting export go-routine. I will give that a shot in the next commit. If that doesn't look good, I will switch to the suggested way.

Reply:
Found that using mutex for this purpose with select could become tricky. Followed the suggestion of using multiple signals and handled the signal in the export goroutine based on number of expected poll cycles before starting export. Used unbuffered channel, let me know if you have an opinion on buffered channel. Found buffered channel to be risky for this purpose.

Reply:
> reading the code, it seems that the name is not appropriate. You do not send a signal when polling is done, you send a signal when the exporter needs to run.

As I changed the signal mechanism to send multiple signals to export routine, I kept the same name pollDone.